app/com/linkedin/drelephant/tuning/hbt/MRApplicationData.java

```diff
@@ -0,0 +1,740 @@
+      float previousSortSpill) {
+    int newBufferSize = 0;
+    float newSpillPercentage = 0.0f;
+    if (ratioOfDiskSpillsToOutputRecords >= RATIO_OF_DISK_SPILL_TO_OUTPUT_RECORDS_THRESHOLD_FIRST.getValue()) {
+      if (previousSortSpill <= SORT_SPILL_THRESHOLD_FIRST.getValue()) {
+        newSpillPercentage = previousSortSpill + SORT_SPILL_INCREASE.getValue();
+        newBufferSize = (int) (previousBufferSize * BUFFER_SIZE_INCREASE_FIRST.getValue());
+      } else {
+        newBufferSize = (int) (previousBufferSize * BUFFER_SIZE_INCREASE_SECOND.getValue());
+      }
+    } else if (ratioOfDiskSpillsToOutputRecords >= RATIO_OF_DISK_SPILL_TO_OUTPUT_RECORDS_THRESHOLD_SECOND.getValue()) {
+      if (previousSortSpill <= SORT_SPILL_THRESHOLD_FIRST.getValue()) {
+        newSpillPercentage = previousSortSpill + SORT_SPILL_INCREASE.getValue();
+        newBufferSize = (int) (previousBufferSize * BUFFER_SIZE_INCREASE.getValue());
+      } else {
+        newBufferSize = (int) (previousBufferSize * BUFFER_SIZE_INCREASE_FIRST.getValue());
+      }
+    }
+    logDebuggingStatement(" Previous Buffer " + previousBufferSize, " Previous Split " + previousSortSpill,
+        "Ratio of disk spills to output records " + ratioOfDiskSpillsToOutputRecords,
+        "New Buffer Size " + newBufferSize * 1.0, " New Buffer Percentage " + newSpillPercentage);
+    if (newBufferSize > 0) {
+      suggestedParameter.put(SORT_BUFFER_HADOOP_CONF.getValue(), newBufferSize * 1.0);
+      suggestedParameter.put(SORT_SPILL_HADOOP_CONF.getValue(), newSpillPercentage * 1.0);
+      modifyMapperMemory();
+    }
+  }
+
+  private int getPreviousBufferSize() {
+    int previousBufferSize = 0;
+    if (appliedParameter.containsKey(SORT_BUFFER.getValue())) {
+      previousBufferSize = appliedParameter.get(SORT_BUFFER.getValue()).intValue();
+    }
+    return previousBufferSize;
+  }
+
+  private float getPreviousSpill() {
+    float previousSpill = 0.0f;
+    if (appliedParameter.containsKey(SORT_SPILL.getValue())) {
+      previousSpill = appliedParameter.get(SORT_SPILL.getValue()).floatValue();
+    }
+    return previousSpill;
+  }
+
+  private float getRatioOfDiskSpillsToOutputRecords(Map<String, String> heuristicsResults) {
+    String ratioOfSpillRecordsToOutputRecordsValue = heuristicsResults.get(
+        MapperSpillHeuristic.class.getCanonicalName() + "_" + RATIO_OF_SPILLED_RECORDS_TO_OUTPUT_RECORDS.getValue());
+    float ratioOfDiskSpillsToOutputRecords = 0.0f;
+    try {
+      if (ratioOfSpillRecordsToOutputRecordsValue != null) {
+        ratioOfDiskSpillsToOutputRecords = Float.parseFloat(ratioOfSpillRecordsToOutputRecordsValue);
+      }
+    } catch (NumberFormatException e) {
+      logger.error("Error parsing ration of spill records to output records ", e);
+    }
+    return ratioOfDiskSpillsToOutputRecords;
+  }
+
+  private void modifyMapperMemory() {
```

Comment:
This method is called by `calculateAndUpdateNewValuesForBuffer` which is called before suggesting Mapper Memory in case both Mapper Memory and Mapper Spill heuristics are failed. So if the mapper memory suggested in accordance to the Buffer size by method **modifyMapperMemory** can be more than the Mapper Memory suggested by method **processForMemory** for Mapper. This would result in wrong suggestion for the Job.

Reply:
In processHeuristics , first modify mapper memory is called if both heuristics are failed . Then spill heuristics called , if the buffer size changes then again modify mapper is called.  I am not sure what exactly you mean , by wrong suggestion

Reply:
My thought was that since the order to suggest for Heuristics is not fixed, there are just conditional statements, it can be a possibility the MapperSpill heuristic is processed first than the Mapper Memory.

Reply:
Changed , in case Mapper Spill suggested first than Mapper memory

Reply:
As discussed offline, Mapper Memory heuristics will be processed before the Mapper Spill.

Reply:
Resolved